### PR TITLE
Add chain on _type parameter

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/4_3_0/1772-allow-chain-on-type.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+type: add
+issue: 1772
+title: "The JPA server now allows chained searches on the `_type` parameter. For example, the following
+   could be used to find all Encounters with a context of type Group: `Encounter?context._type=Group`."
```
Comment (LishaZ) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/4_3_0/1772-allow-chain-on-type.yaml]: Nice that you have this changelog! 
I didn't see `context` in `Encounter`: https://www.hl7.org/fhir/encounter.html. Perhaps you could use `subject` instead?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/predicate/PredicateBuilderReference.java [hunk 2]
```diff
@@ -273,20 +304,43 @@ private Predicate addPredicateReferenceWithChain(String theResourceName, String
...
+			Class<? extends IBaseResource> wantedType;
+			try {
+				wantedType = myContext.getResourceDefinition(typeValue).getImplementingClass();
+			} catch (DataFormatException e) {
+				throw newInvalidResourceTypeException(typeValue);
+			}
+			if (!resourceTypes.contains(wantedType)) {
+				String searchParamName = theResourceName + ":" + theParamName;
+				String msg = myContext.getLocalizer().getMessage(PredicateBuilderReference.class, "invalidTargetTypeForChain", typeValue, searchParamName);
+				throw new InvalidRequestException(msg);
```
Comment (LishaZ) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/predicate/PredicateBuilderReference.java]: Maybe you can refactor this into a method like you did for `newInvalidResourceTypeException` so it's a bit easier to read

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java [hunk 3]
```diff
@@ -316,6 +316,67 @@ public void testChainWithMultipleTypePossibilities() {
...
+		map = new SearchParameterMap();
+		map.setLoadSynchronous(true);
+		map.add(Encounter.SP_SUBJECT, new ReferenceParam("subject", "Patient").setChain("_type"));
+		results = myEncounterDao.search(map);
+		ids = toUnqualifiedVersionlessIdValues(results);
+		assertThat(ids, hasItems(enc1Id));
+
+		map = new SearchParameterMap();
+		map.setLoadSynchronous(true);
+		map.add(Encounter.SP_SUBJECT, new ReferenceParam("subject", "Group").setChain("_type"));
```
Comment (LishaZ) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java]: I assume this means `subject._type=Group`

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/4_3_0/1772-allow-chain-on-type.yaml (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/4_3_0/1772-allow-chain-on-type.yaml]: Whoops! Will fix.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/predicate/PredicateBuilderReference.java (hunk 2 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/predicate/PredicateBuilderReference.java]: Makes sense, done

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java]: Yup!

**Pull request merged**